src/TorchSharp/AutogradFunction.cs

```diff
@@ -0,0 +1,502 @@
+                /// </summary>
+                internal void save_variables()
+                {
+                    _savedVariables.Clear();
+                    lock (_node) {
+                        foreach (var tensor in _tensorsToSave)
+                            _savedVariables.Add(tensor is null || tensor.IsInvalid ? new() : new(tensor, _node.handle));
+                    }
+                }
+
+                internal List<Tensor> DirtyTensors => _dirtyTensors;
+                internal List<Tensor> NonDifferentiableTensors => _nonDifferentiableTensors;
+
+                public void Dispose()
+                {
+                    Dispose(true);
+                    GC.SuppressFinalize(this);
+                }
+
+                protected virtual void Dispose(bool disposing)
+                {
+                    if (disposing) {
+                        _savedVariables?.ForEach(v => v.Dispose());
+                        _savedVariables = null;
+                    }
+                }
+            }
+
+            /// <summary>
+            /// To use custom autograd operations, implement a Function subclass implementing tthe
+            /// forward and backward functions:
+            /// `forward` can take as many arguments as you want, passed an object[], and will return either
+            /// a list of tensors or a single tensor (depending which subclass you inherit).
+            /// Use of any direct Variable arguments will be registered in the graph but no vectors/sets or any
+            /// other data structures will be traversed. You can pass null tensors as one of the arguments
+            /// and it will be registered as a variable in the graph if the argument has a
+            /// value. It should take a pointer to `torch::autograd::AutogradContext` as the
+            /// first argument. Tensors can be saved in the `ctx` using `ctx->save_for_backward`
+            /// (see `torch::autograd::AutogradContext::save_for_backward`) and other data
+            /// can be saved using the `ctx->save_data` function (see `torch::autograd::AutogradContext::save_data`)
+            /// in the form of `(string, object)` pairs.
+            /// Variables saved in `forward` can be accessed with `ctx->get_saved_variables` (see
+            /// `torch::autograd::AutogradContext::get_saved_variables`) and other saved
+            /// data can be accessed using `ctx->get_data()`.
+            /// </summary>
+            public abstract class Function<T> where T : Function<T>, new()
+            {
+                internal static T Instance { get; set; } = new();
+
+                /// <summary>
+                /// When calling the function, the user should call Function.apply and not the forward function, so that
+                /// the computation graph is built correctly.
+                /// </summary>
+                public static List<Tensor> apply_internal(params object[] vars)
```

Comment:
What else than Tensor could the parameters be?

Reply:
An int, double, floatt.  
See here an example: https://pytorch.org/cppdocs/api/structtorch_1_1autograd_1_1_function.html

Reply:
Also here: https://pytorch.org/tutorials/advanced/cpp_autograd.html
For example, the MulConstant